### ISSUE TITLE
Proposal: Improve widget selection in the inspector

### ISSUE THREAD

**Issue by CodemateTechAgency** [flutter/devtools#3489]
# KR1

## Background

> “The inspector allows the user to search the summary tree and the details tree by keyword. The inspector searches the part of the details tree that has been loaded by default with an option to search the full details tree enabled by KR 2.”

## Proposal

- Add bordered headers for both widget tree and explorer/details areas to isolate browsing/filtering related actions and separate them from the action toolbar above. Tab bar look should reflect the rest of the DevTools (without the grey background).
- Add search functionality to the Widget Tree header. Move the refresh widget tree to the header area.
  - Search can target different contexts based on the user input. (Widget/Details/All etc.)
- Rename “Details tree” to “Widget Details Tree” to pair it better with the Widget Tree.

**Question**: Should we now move the select widget mode to the button group on the right?

### Before

![Screenshot 2021-11-05 at 10 11 30](https://user-images.githubusercontent.com/82218649/140737406-09100e5e-eefb-47f1-a10f-2349af0be678.png)

### After

![Screenshot 2021-11-05 at 10 12 17](https://user-images.githubusercontent.com/82218649/140737603-b9b9eca9-c154-44f7-b117-00e4c00b1338.png)

![Screenshot 2021-11-05 at 10 12 09](https://user-images.githubusercontent.com/82218649/140737648-d8056a42-d59b-4891-aef7-e2df6540182f.png)

### Search field variations

![Screenshot 2021-11-08 at 10 50 32](https://user-images.githubusercontent.com/82218649/140737678-c2fb72da-f6a5-4156-8985-4363e55ff5e1.png)

Normal state **a)**, active search fields with **b)** and without **c)** search options (search from widget names, details, all, etc). Floating version **d)** for narrow spaces 

---

# KR2

## Background

> “The inspector provides a UI affordance to show the complete detail tree. This is essentially a shortcut for performing the following steps: 1) select the root widget in the summary tree and 2) click the details tree tab. Progressive loading might be needed for very large trees.“

Additional input and background information is welcome. Assumption here is that KR2 will support KR1 with more detailed search and navigation. As KR1 can also include “Widget Details Tree” as a search target, this proposal is to make the navigation tree more discoverable.

## Proposal

- Add breadcrumb navigation atop of the details tree for easier navigation and to make root level always accessible and easily discoverable.

### Before

![Screenshot 2021-11-08 at 14 00 24](https://user-images.githubusercontent.com/82218649/140738513-d86009c0-c1eb-4779-926d-8940809fec1b.png)

### After

![Screenshot 2021-11-08 at 13 58 42](https://user-images.githubusercontent.com/82218649/140738334-57ac9dc9-d461-46bf-8767-2b2075c185d2.png)

---

Any input is greatly appreciated!

cc @InMatrix @csells @jacob314 @kenzieschmoll

**Comment by kenzieschmoll** [comment]
This looks great. For KR1, I like the idea of clicking the search icon for option A animating to search field option B. And as long as there is an active search, the search field would be present - if the search field is empty and it loses focus, the search field would animate closed back to option A.

LGTM to new tab UI and titles, new placement refresh button, and breadcrumb navigator in KR2

**Comment by InMatrix** [comment]
> Add bordered headers for both widget tree and explorer/details areas ...

I like the headers. Since we have them now, can we add tooltips to explain what we're showing in each area? For example, a tooltip for the "Widget Tree" header can say, "Widgets used directly to compose your app's UI". 

>Normal state a), active search fields with b) and without c) search options (search from widget names, details, all, etc). Floating version d) for narrow spaces

I like the search bar with a scope setting (option B), and I think that setting should persist between search sessions until the user changes it explicitly. It's debatable whether the default search scope should be just the user widgets on the left or all widgets (those on the right). 

My question is what exactly happens when the scope is set to "all"? Does the details tree automatically expand to show everything and then matches the search query? 

> Question: Should we now move the select widget mode to the button group on the right?

I don't think that's necessary? Let's revisit this when we discuss proposals for KR3. 

> Add breadcrumb navigation atop of the details tree for easier navigation and to make root level always accessible and easily discoverable.

I like this idea. Could you clarify whether the visible upper levels in the breadcrumb nav are the higher-level user widgets in the tree or the immediate parents of the selected widget in the details tree? It feels like the former based on the mock, but I'm not sure.

**Comment by CM-Olli** [comment]
> My question is what exactly happens when the scope is set to "all"? Does the details tree automatically expand to show everything and then matches the search query?

We would suggest that the search field has 3 scopes:

- Widget - Search "Widget tree" by name and description (this would be the default scope)
- Details - Search "Details tree" properties (key/value)
- All - Enables both scopes described above

Here's a visual presentation of these 3 scopes to better highlight the proposed functionality:

![Screenshot 2021-11-10 at 8 30 57](https://user-images.githubusercontent.com/86104144/141092847-51261cef-75ec-4b50-ba13-b5855eab2529.png)
Widget: Get matches from widget tree side, but not from details

![Screenshot 2021-11-10 at 8 31 02](https://user-images.githubusercontent.com/86104144/141092873-1872e3fd-bc7b-4bc8-bad3-0879f4b56500.png)
Details: Highlighting widgets that have matches in its details.  


![Screenshot 2021-11-10 at 8 31 18](https://user-images.githubusercontent.com/86104144/141092981-291f1f6a-3cd6-437e-b7b7-541e2fbf7b55.png)
![Screenshot 2021-11-10 at 8 31 10](https://user-images.githubusercontent.com/86104144/141093033-0b43bfe8-e72b-41f0-9bbe-2524ace0773b.png)
All: Highlighting widgets, that has matches in its details and name and description 

Does this sound like a reasonable approach?

Related to "Details" -scope: One question we initially had was the “Details tree” search use cases. Could you clarify the real-world use cases of targeting the search on “Details tree”?

> Could you clarify whether the visible upper levels in the breadcrumb nav are the higher-level user widgets in the tree or the immediate parents of the selected widget in the details tree? It feels like the former based on the mock, but I'm not sure.

Former, we propose that the breadcrumb visibility should reflect the selected widget on a “Widget tree”.

**Comment by msavela** [comment]
There are obviously some performance concerns with the "Details" -scope and searching the whole "Widget Details Tree". In some cases it can take really long time to finish.

I did some benchmarks around this topic and it took around 15-20 secs when performing "Details" -scope search while debugging DevTools itself. The good news is that the search could be performed asynchronously and matches can be updated as the search progresses.

This might not be a problem with smaller projects, however this could reduce the user experience with more complex widget tree.

**Comment by InMatrix** [comment]
I agree that the "Widget" search scope should be the default. It also matches the location of the search bar. 

Searching the whole details tree is not a common use case, but it might be important to Flutter framework engineers and package developers.  @Hixie I think you mentioned that you wanted this feature, could you elaborate on your use case? 

I'm not sure if we need the "search all" scope, since the user might either just care about the widgets directly used in their project or the full details which include the user widgets which would include the widgets in the summary tree.

Regarding search performance, @kenzieschmoll is there room for optimization here? @msavela It would be helpful to see a demo to understand how well incremental updates of search results work.

**Comment by kenzieschmoll** [comment]
> Regarding search performance, @kenzieschmoll is there room for optimization here?

Generally, we could try to do things like cache search results, not search until the string is 2 or 3 chars long, add a delay between keystrokes before triggering a search, or make sure we cancel any search operation if a new one is started.

**Comment by msavela** [comment]
Below you can find brief demonstration of both Widget and Details search in action:

**Widget search**
![search_widget](https://user-images.githubusercontent.com/1315762/141812612-813f92ba-6642-4ef1-a690-51836e2084a9.gif)

**Details search**
![search_details](https://user-images.githubusercontent.com/1315762/141812662-762e3d22-bd82-4246-9277-4a8dd729c294.gif)

^ Above you can see how the incremental search results would work

**Comment by InMatrix** [comment]
Re: performance concern in searching the details tree. Can we do some time-boxed investigation into performance improvements? I followed up on the use case with @Hixie on Discord ([link to the messages](https://discord.com/channels/608014603317936148/608020249119555594/909886751248646184)). It seems that we can focus on searching widget names. @Hixie also suggested representing the widget tree data differently to improve performance.

**Comment by msavela** [comment]
I agree that this would be a good approach. We should first focus on Widget -search and rest of the proposed changes. Once that implementation is finished and PR is open we can time-box this performance investigation together for separated work item and possible PR.

**Comment by CM-Olli** [comment]
### Suggestions for KR3:

**Step 1:**
Change icon to the same one as in Chrome
Tool tip to select widget mode button explaining the functionalities (hold & drag selection etc) + link to docs

![tooltip](https://user-images.githubusercontent.com/86104144/145818761-fe4fd0ca-a291-4333-bfd2-1188974f42a8.png)



- Similar toggle functionality like in Chrome (toggle)
- Always show the select button in the view, when select widget mode is active. 
  (maybe add a minimize button to get it out of the way if needed, or an animated slide down effect when selecting a new widget) 
- Possibility to exit the select mode from the emulator.

![toggle](https://user-images.githubusercontent.com/86104144/145819168-59c5b281-03c3-4fc6-baf6-bf259459e5a5.png)


Long press to drag and select widgets in the device like in the emulator.
Is this already working on a physical device? At least with holding the mouse button you can drag and select around the emulator. If not, then the long press could work similarily.




**Few ideas to improve the app side of things:**
- A breadcrumb to the bottom, so user could move to a parent widget when clicking on area containin multiple widgets. It could show all the widgets and be scrollable.
- Add a scrollwheel-like layout to the selected widget label to move up and down the tree.
- ...or a mompact version of the previous, with current widget name + arrows for navigation

![labels](https://user-images.githubusercontent.com/86104144/145820290-c6112b64-4a35-4762-ad40-7f704a0154f8.png)


We could show measurements (size, margins etc.) of the selected widget here as well


**Something else loosely related:**
Rename select mode to something else (e.g. “Developer Overlay”) and add few functionalities: 
(screen size for desktop, toggle dark mode, language selector, locale, test direction, text size, accessibility options, platform mechanics)


Pairing the select mode w. other toggles to focus them on a certain widget 
(ie repaint only one widget, guides & margins for the selected etc etc)

**Comment by CM-Olli** [comment]
> State 3: A widget has been selected and confirmed
> The "Select Widget" bar shows up at the bottom of the screen as an overlay. _We might want to add some kind of onboarding UI to demonstrate how to use this feature._

Would the info on the devtools tooltip be enough here? If the behaviour would be changed to some of the following ideas (more chrome-like), would this feel “familiar” without additional onboarding? Nonetheless, the bottom of the screen has room for info-icons, help texts etc.


> The main UX question is about State 2 & 3. The on-screen button for selecting another widget is currently hard to discover and awkwardly placed. Your proposal of turning it a bottom bar could be helpful, since it's more noticeable and we can use the extra space to show some help content about the feature. The downside is that when a widget underneath the bar is selected, it might look awkward. _Maybe if that happens, the bar can snap to another side/corner of the screen or float somewhere else?_ I'd appreciate some UI explorations for that situation.

One thought on this was that when entering the state 1 & 4, the bottom bar could be minimized via clicking the minimize icon or automatically (with an animation instead of just disappearing, so users know what is going on) and bringing the bar back up at state 3. That alone probably won’t solve all problems with the widgets underneath, although the selection would be visible under the semi-transparent bottom bar. 

Adding the option to relocate the bar, either dragging manually, or automatically when selecting the widgets that are underneath it would work nicely.


> An alternative to the long-press method used in State 2 is to allow the user to click and select another widget until they explicitly exit the "Select Widget Mode" from DevTools. In other words, user input events will continue getting captured by DevTools in State 3. I'm not sure what might be the downsides of doing this instead. _I'd love to learn your thoughts._

Maybe instead of controlling this from devtools alone, we could use the select icon in app view for this. Instead of “single” select, entering the select widget states (1&4) you’d be able to select widgets until toggled off manually from the same button in the app screen. 

At first it feels like like a duplicate functionality for toggling the the select widget mode on/off in devtools, but this would be in a more convenient location (and controllable without devtools open), separate from exiting the mode and would allow utilizing the bottom bar for other stuff as well, like info texts.

I guess it also depends on the most common use case - is this used to select and go through multiple widgets in one view / few across the whole application.


> Another open question your proposal hasn't addressed is whether DevTools should let users leverage the hover state when the app is running in a browser or a desktop environment. We don't want the user to relearn how to use the tool in those environments, but there might be opportunities to provide extra convenience since the capability is available. _What do you think?_

In our discussions, the main issues with select widget mode were the lack of instructions and especially the lack of control over your selection, so we focused our suggestions to tackle those first.

Also, we thought that it would be a good idea to bring the functionalities a bit more towards Chrome, since that is something that many are familiar with (icons, breadcrumbs, behavior etc.) There, entering the select state utilizes hover and acts like click ‘n hold + drag currently on the devtools’ app screen.

Like you mentioned, things should be consistent across the way, so the first thing that comes to mind is adding supporting features for hover state. Tooltips showing the name of the widget hovering above is a first thought.  Or what would you think, would it be confusing that selecting the widget on the desktop would always behave like Chrome (hover-highlighting widget under cursor, click to select it), but on the device, long press & drag would do the same (which i suppose would be expected anyway), and you really wouldn’t miss anything important.

Feel free to add your thoughts too, @msavela @tomgilder

**Comment by Hixie** [comment]
Drive-by comment: One thing we should be careful of when designing these UIs is that every pixel on the screen might be interesting. For example, in the screenshot above, consider how the user would indicate they want to inspect a widget that's rendered below the floating action button, at the bottom of the screen (where the inspector buttons are currently).

**Comment by InMatrix** [comment]
> Would the info on the devtools tooltip be enough here?

I think it won't be enough because 1) the emulator and device could be spatially far away from the DevTools window depending on the developer's setup, and 2) this interaction is hard to explain through text in a tooltip. Ideally, this information should be delivered when the user has entered State 3. The current workflow often confuses users when they couldn't just click on another widget and have it selected after their initial selection. 

> If the behaviour would be changed to some of the following ideas (more chrome-like), would this feel “familiar” without additional onboarding?  

Which chrome-like idea are you referring to? It's quite possible to emulate the Chrome widget selection experience when the Flutter app is running in the web browser or a desktop environment, and I'd agree that it's probably unnecessary to explain how the feature works. But it's hard to do that for a mobile device or emulator due to lack of the hover event. 

> One thought on this was that when entering the state 1 & 4, the bottom bar could be minimized via clicking the minimize icon or automatically (with an animation instead of just disappearing, so users know what is going on) and bringing the bar back up at state 3. That alone probably won’t solve all problems with the widgets underneath, although the selection would be visible under the semi-transparent bottom bar.
> 
> Adding the option to relocate the bar, either dragging manually, or automatically when selecting the widgets that are underneath it would work nicely.

That's exactly what I was thinking and it should address @Hixie's concern.

> Maybe instead of controlling this from devtools alone, we could use the select icon in app view for this. Instead of “single” select, entering the select widget states (1&4) you’d be able to select widgets until toggled off manually from the same button in the app screen.

I think this might be preferable on mobile, since it's hard to select the widget you want precisely with a single click, given the nested nature of Flutter widgets. I don't remember the rationale for the current "single select" interaction. @kenzieschmoll Do you know? 

> Or what would you think, would it be confusing that selecting the widget on the desktop would always behave like Chrome (hover-highlighting widget under cursor, click to select it), but on the device, long press & drag would do the same (which i suppose would be expected anyway), and you really wouldn’t miss anything important.

I think we currently lack a way to teach the user that long press & drag is equivalent to hover over on desktop and in Chrome. The current on-screen select button is a workaround for lacking that instruction. I think we have two ways to make the workflow more intuitive:
1. Always go back to State 1 (ready to select) after selecting a widget, which will make both long press & drag and the current "select another widget" button on the app screen unnecessary. The slight downside is that the user will need to exit the selection mode explicitly before they can interact with the app again. 
1. Add instructions about how to select another widget either through long press & drag or the on-screen select button.

**Comment by kenzieschmoll** [comment]
> I don't remember the rationale for the current "single select" interaction. @kenzieschmoll Do you know?

I do not know. This has been like this for as long as I've been around.

**Comment by aednlaxer** [comment]
I've created a draft PR with experimental widget detail search: #3552 Kindly ask you to check how performant it is. As mentioned above, search may take several seconds when widget tree is huge (for example devtools is debugging itself)

**Comment by InMatrix** [comment]
@aednlaxer Thanks for the update. I requested @kenzieschmoll to look at the performance of your PR.

For the search single-select vs. continuous-select question, we probably won't be able to resolve it until @jacob314, who made the original implementation of this functionality, is back in January.

**Comment by InMatrix** [comment]
Updated: we're likely to prototype an alternate widget selection workflow this quarter, so the user won't exit the active selection mode until they explicitly do so using the selection mode toggle button on the inspector page. There is a separate bug tracking that work: https://github.com/flutter/devtools/issues/3567. The UX design work will be on hold until we have an opportunity to examine the alternate workflow in action.
